### PR TITLE
Make `ERR_str_reasons` in `err.c` consistent again with `err.h`

### PR DESCRIPTION
**Pull request by DDvO**
For various generic reason codes, their string conversion goes wrong, e.g.: 
```
    # 80B26552927F0000:error:1E8C0102:HTTP routines:foo:reason(786690):test/http_test.c:424:
    # 80B26552927F0000:error:1E880106:HTTP routines:foo:reason(524550):test/http_test.c:424:
```

I found that this is because some entries defined in `include/openssl/err.h{,in}` are missing from `ERR_str_reasons` in `crypto/err/err.c`
and that their order easily goes wrong there because they need to be sorted by _their numerical reason code **including** any flags_ such as `ERR_RFLAG_COMMON`. 

After fixing the entries, the above becomes:
```
    # 8032C819DE7F0000:error:1E8C0102:HTTP routines:foo:passed a null parameter:test/http_test.c:424:
    # 8032C819DE7F0000:error:1E880106:HTTP routines:foo:passed an invalid argument:test/http_test.c:424:
```

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
> flags become part of the index where they shouldn't.

Right - I'll provide a fixup that rectifies this, which will make `ERR_str_reasons[]` better maintainable.

**Comment by DDvO** [comment]
> I'll provide a fixup that rectifies this, which will make `ERR_str_reasons[]` better maintainable.

Done.

After some pitfalls, seems to be fine now.

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Actually, all that should be needed is to add the missing strings in `ERR_str_reasons`.  Numeric order should not matter, since they are indexed by an lhash (`int_error_hash`)!

Now, I did try printing the two codes, and here's what I got:

``` console
$ ./util/wrap.pl apps/openssl errstr 1E8C0102
error:1E8C0102:HTTP routines::passed a null parameter
$ ./util/wrap.pl apps/openssl errstr 1E880106
error:1E880106:HTTP routines::reason(524550)
```

And indeed, there is no text string for `ERR_R_PASSED_INVALID_ARGUMENT` in `crypto/err/err.c`.

Why you got no text reason for 1E8C0102, I cannot tell.

**Comment by levitte** [comment]
Regarding the flags, I'm backing off what I said earlier.  Considering that we're using the same reason macros everywhere, it really doesn't matter if we peal off the flags or not, as they should be the same anyway...  that was a red herring.

**Comment by levitte** [comment]
I just submitted #17069, which only adds the missing reason string.  To be noted that it's missing all the way back to 1.1.1.

The rest of this PR should probably be discussed a little further, because as noted in the modified test, it changes a previous behavior, something I didn't think of initially.  That's probably not well suited for 3.0 or 1.1.1.

File: crypto/err/err.c [hunk 1]
```diff
@@ -92,17 +93,16 @@ static ERR_STRING_DATA ERR_str_reasons[] = {
     {ERR_R_DSA_LIB, "DSA lib"},
     {ERR_R_X509_LIB, "X509 lib"},
     {ERR_R_ASN1_LIB, "ASN1 lib"},
+    {ERR_R_CRYPTO_LIB, "crypto lib"},
```
Comment (levitte) [inline crypto/err/err.c:96]: ?

File: crypto/err/err.c [hunk 2]
```diff
@@ -92,17 +93,16 @@ static ERR_STRING_DATA ERR_str_reasons[] = {
...
     {ERR_R_EC_LIB, "EC lib"},
     {ERR_R_BIO_LIB, "BIO lib"},
     {ERR_R_PKCS7_LIB, "PKCS7 lib"},
     {ERR_R_X509V3_LIB, "X509V3 lib"},
-    {ERR_R_ENGINE_LIB, "ENGINE lib"},
+    {ERR_R_ENGINE_LIB, "engine lib"},
```
Comment (levitte) [inline crypto/err/err.c:101]: ?

File: crypto/err/err.c [hunk 3]
```diff
@@ -92,17 +93,16 @@ static ERR_STRING_DATA ERR_str_reasons[] = {
...
     {ERR_R_UI_LIB, "UI lib"},
-    {ERR_R_OSSL_STORE_LIB, "STORE lib"},
     {ERR_R_ECDSA_LIB, "ECDSA lib"},
-
-    {ERR_R_NESTED_ASN1_ERROR, "nested asn1 error"},
-    {ERR_R_MISSING_ASN1_EOS, "missing asn1 eos"},
+    {ERR_R_OSSL_STORE_LIB, "store lib"},
+    {ERR_R_OSSL_DECODER_LIB, "decoder lib"},
```
Comment (levitte) [inline crypto/err/err.c:105]: ?

**Review by levitte — CHANGES_REQUESTED** [review]
Why are you lowercasing the library names?  They are uppercase to match the API prefixes.
(and yes, one might argue that "STORE lib" should be "OSSL_STORE lib")

**Comment by DDvO** [comment]
Yes, the comment on the indexing/order was misleading; it works regardless.
I've pushed a fixup that no more peels off the flags from the list entries, which simplifies things again.
Anyway I made sure the entries in  `err.c` are in the same order as those in `err.h.in`, which should help maintenance.

**Comment by DDvO** [comment]
Indeed the problem was that some entries were missing - but not only `ERR_R_PASSED_INVALID_ARGUMENT`,
 also `ERR_R_CRYPTO_LIB` and `ERR_R_OSSL_DECODER_LIB`, which I added.
 
`test_err*` should better check for completeness of the string outputs.

**Comment by DDvO** [comment]
I added a commit that exempts flags from the fallback decimal reason code printing,
such that, e.g., `reason(524550)` becomes `reason(262)`, which I consider much more readable/helpful.

**Comment by DDvO** [comment]
I also added a tweak to `02-test_errstr.t` 
such that error codes are printed in hex (rather than decimal) format,
which also improves readability (of the verbose output).

File: crypto/err/err.c (hunk 1 above)
Comment (DDvO) [reply crypto/err/err.c:96]: I found `CRYPTO` misleading because `crypto` is not a letter-wise abbreviation, while most others, including `RSA`, are.

File: crypto/err/err.c (hunk 2 above)
Comment (DDvO) [reply crypto/err/err.c:101]: Same reason as above

File: crypto/err/err.c (hunk 3 above)
Comment (DDvO) [reply crypto/err/err.c:105]: `decoder lib` was missing so far and I moved the two ASN.1 error strings to where they below.
Regarding upper vs. lower case see above.

File: crypto/err/err.c (hunk 1 above)
Comment (DDvO) [reply crypto/err/err.c:96]: And already now we do not write `LIB`, but `lib`, for the same reason.

File: crypto/err/err.c (hunk 3 above)
Comment (DDvO) [reply crypto/err/err.c:105]: Ah, right, `OSSL ` was missing. Added in both cases.

File: crypto/err/err.c (hunk 1 above)
Comment (levitte) [reply crypto/err/err.c:96]: Whether it's an abbreviation or not doesn't matter.  There *is* the idea of a `CRYPTO` library (set of functions)
(although to be frank, we have "CRYPTO" functions that are named differently, so I can see the confusion with those)

File: crypto/err/err.c (hunk 2 above)
Comment (levitte) [reply crypto/err/err.c:101]: The `ENGINE` library is the set of functions with the `ENGINE` (not `engine` prefix).

File: crypto/err/err.c [hunk 4]
```diff
@@ -92,17 +93,16 @@ static ERR_STRING_DATA ERR_str_reasons[] = {
...
-    {ERR_R_ENGINE_LIB, "ENGINE lib"},
+    {ERR_R_ENGINE_LIB, "engine lib"},
     {ERR_R_UI_LIB, "UI lib"},
-    {ERR_R_OSSL_STORE_LIB, "STORE lib"},
     {ERR_R_ECDSA_LIB, "ECDSA lib"},
-
-    {ERR_R_NESTED_ASN1_ERROR, "nested asn1 error"},
-    {ERR_R_MISSING_ASN1_EOS, "missing asn1 eos"},
+    {ERR_R_OSSL_STORE_LIB, "OSSL store lib"},
+    {ERR_R_OSSL_DECODER_LIB, "OSSL decoder lib"},
```
Comment (levitte) [inline crypto/err/err.c:105]: Should be "OSSL_STORE lib" and "OSSL_DECODER lib" to be consistent with the others

File: crypto/err/err.c (hunk 1 above)
Comment (DDvO) [reply crypto/err/err.c]: All right, reverted to the UPPERCASE names.

File: crypto/err/err.c (hunk 4 above)
Comment (DDvO) [reply crypto/err/err.c]: I see, done.

File: include/openssl/err.h.in [hunk 5]
```diff
@@ -337,6 +337,7 @@ static ossl_unused ossl_inline int ERR_COMMON_ERROR(unsigned long errcode)
 # define ERR_R_OSSL_DECODER_LIB (ERR_LIB_OSSL_DECODER/* 60 */ | ERR_RFLAG_COMMON)
 
 /* Other common error codes, range 256..2^ERR_RFLAGS_OFFSET-1 */
+/* NOTE: MUST keep ERR_str_reasons in crypto/err/err.c in sync with this list */
```
Comment (levitte) [inline include/openssl/err.h.in:340]: This is uninteresting for the public (this is a public header).  I see no reason for it to be here.

File: crypto/err/err.c [hunk 6]
```diff
@@ -80,6 +80,7 @@ static ERR_STRING_DATA ERR_str_libraries[] = {
     {0, NULL},
 };
 
+/* NOTE: MUST put here all reasons from include/openssl/err.h.in */
```
Comment (levitte) [inline crypto/err/err.c:83]: May I suggest something like this?

``` C
/* For maintainability, please keep all reasons in the same order as in include/openssl/err.h.in */
```

(since there's no MUST, technically speaking)

File: include/openssl/err.h.in (hunk 5 above)
Comment (DDvO) [reply include/openssl/err.h.in]: Good point; removed this.

File: crypto/err/err.c (hunk 6 above)
Comment (DDvO) [reply crypto/err/err.c]: Well, at least all the reasons SHOULD be present. Changed to:
```
/*
 * Should make sure that all ERR_R_ reasons defined in include/openssl/err.h.in
 * are listed.  For maintainability, please keep all reasons in the same order.
 */
```

File: crypto/err/err.c [hunk 7]
```diff
@@ -527,7 +531,8 @@ void ossl_err_string_int(unsigned long e, const char *func,
     }
 #endif
     if (rs == NULL) {
-        BIO_snprintf(rsbuf, sizeof(rsbuf), "reason(%lu)", r);
+        BIO_snprintf(rsbuf, sizeof(rsbuf), "reason(%lu)",
+                     r & ~(ERR_RFLAGS_MASK << ERR_RFLAGS_OFFSET));
```
Comment (levitte) [inline crypto/err/err.c:535]: I would approve everything of this PR, apart for this particular change.  It's not strictly necessary with regards to missing string, which is the actual bug you're trying to fix.

I don't mean to say that this isn't a change worth thinking about, but it might not be suitable for 3.0...  so I would suggest moving this change to a different PR.

Comment (DDvO) [reply crypto/err/err.c]: Pleased to hear.
Moved that part to #17072.

**Comment by DDvO** [comment]
> Why you got no text reason for 1E8C0102, I cannot tell.

I cannot reproduce this part of the issue (regarding `passed a null parameter`) myself.
Anyway, for the records, this is the output of all reasons defined in `err.h.in` before:
```
801BB4FE0C7F0000:error:01880002:bignum routines:test_reason_strings:system lib:test/errtest.c:339:
801BB4FE0C7F0000:error:01880003:bignum routines:test_reason_strings:BN lib:test/errtest.c:340:
801BB4FE0C7F0000:error:01880004:bignum routines:test_reason_strings:RSA lib:test/errtest.c:341:
801BB4FE0C7F0000:error:01880005:bignum routines:test_reason_strings:DH lib:test/errtest.c:342:
801BB4FE0C7F0000:error:01880006:bignum routines:test_reason_strings:EVP lib:test/errtest.c:343:
801BB4FE0C7F0000:error:01880007:bignum routines:test_reason_strings:BUF lib:test/errtest.c:344:
801BB4FE0C7F0000:error:01880008:bignum routines:test_reason_strings:OBJ lib:test/errtest.c:345:
801BB4FE0C7F0000:error:01880009:bignum routines:test_reason_strings:PEM lib:test/errtest.c:346:
801BB4FE0C7F0000:error:0188000A:bignum routines:test_reason_strings:DSA lib:test/errtest.c:347:
801BB4FE0C7F0000:error:0188000B:bignum routines:test_reason_strings:X509 lib:test/errtest.c:348:
801BB4FE0C7F0000:error:0188000D:bignum routines:test_reason_strings:ASN1 lib:test/errtest.c:350:
801BB4FE0C7F0000:error:0188000F:bignum routines:test_reason_strings:reason(524303):test/errtest.c:351:
801BB4FE0C7F0000:error:01880010:bignum routines:test_reason_strings:EC lib:test/errtest.c:352:
801BB4FE0C7F0000:error:01880020:bignum routines:test_reason_strings:BIO lib:test/errtest.c:353:
801BB4FE0C7F0000:error:01880021:bignum routines:test_reason_strings:PKCS7 lib:test/errtest.c:354:
801BB4FE0C7F0000:error:01880022:bignum routines:test_reason_strings:X509V3 lib:test/errtest.c:355:
801BB4FE0C7F0000:error:01880026:bignum routines:test_reason_strings:ENGINE lib:test/errtest.c:356:
801BB4FE0C7F0000:error:01880028:bignum routines:test_reason_strings:UI lib:test/errtest.c:357:
801BB4FE0C7F0000:error:0188002A:bignum routines:test_reason_strings:ECDSA lib:test/errtest.c:358:
801BB4FE0C7F0000:error:0188002C:bignum routines:test_reason_strings:STORE lib:test/errtest.c:359:
801BB4FE0C7F0000:error:0188003C:bignum routines:test_reason_strings:reason(524348):test/errtest.c:361:
801BB4FE0C7F0000:error:018C0000:bignum routines:test_reason_strings:fatal:test/errtest.c:362:
801BB4FE0C7F0000:error:018C0100:bignum routines:test_reason_strings:malloc failure:test/errtest.c:363:
801BB4FE0C7F0000:error:018C0101:bignum routines:test_reason_strings:called a function you should not call:test/errtest.c:364:
801BB4FE0C7F0000:error:018C0102:bignum routines:test_reason_strings:passed a null parameter:test/errtest.c:365:
801BB4FE0C7F0000:error:018C0103:bignum routines:test_reason_strings:internal error:test/errtest.c:366:
801BB4FE0C7F0000:error:018C0104:bignum routines:test_reason_strings:called a function that was disabled at compile-time:test/errtest.c:367:
801BB4FE0C7F0000:error:018C0105:bignum routines:test_reason_strings:init fail:test/errtest.c:368:
801BB4FE0C7F0000:error:01880106:bignum routines:test_reason_strings:reason(524550):test/errtest.c:369:
801BB4FE0C7F0000:error:018C0107:bignum routines:test_reason_strings:operation fail:test/errtest.c:371:
801BB4FE0C7F0000:error:018C0108:bignum routines:test_reason_strings:invalid provider functions:test/errtest.c:372:
801BB4FE0C7F0000:error:01880109:bignum routines:test_reason_strings:interrupted or cancelled:test/errtest.c:373:
801BB4FE0C7F0000:error:0188010A:bignum routines:test_reason_strings:nested asn1 error:test/errtest.c:374:
801BB4FE0C7F0000:error:0188010B:bignum routines:test_reason_strings:missing asn1 eos:test/errtest.c:375:
801BB4FE0C7F0000:error:0188010C:bignum routines:test_reason_strings:unsupported:test/errtest.c:376:
801BB4FE0C7F0000:error:0188010D:bignum routines:test_reason_strings:fetch failed:test/errtest.c:377:
801BB4FE0C7F0000:error:0188010E:bignum routines:test_reason_strings:invalid property definition:test/errtest.c:378:
801BB4FE0C7F0000:error:018C010F:bignum routines:test_reason_strings:unable to get read lock:test/errtest.c:379:
801BB4FE0C7F0000:error:018C0110:bignum routines:test_reason_strings:unable to get write lock:test/errtest.c:380:

```
and after the improvements by this PR:
```
808B11CDB37F0000:error:01880002:bignum routines:test_reason_strings:system lib:test/errtest.c:339:
808B11CDB37F0000:error:01880003:bignum routines:test_reason_strings:BN lib:test/errtest.c:340:
808B11CDB37F0000:error:01880004:bignum routines:test_reason_strings:RSA lib:test/errtest.c:341:
808B11CDB37F0000:error:01880005:bignum routines:test_reason_strings:DH lib:test/errtest.c:342:
808B11CDB37F0000:error:01880006:bignum routines:test_reason_strings:EVP lib:test/errtest.c:343:
808B11CDB37F0000:error:01880007:bignum routines:test_reason_strings:BUF lib:test/errtest.c:344:
808B11CDB37F0000:error:01880008:bignum routines:test_reason_strings:OBJ lib:test/errtest.c:345:
808B11CDB37F0000:error:01880009:bignum routines:test_reason_strings:PEM lib:test/errtest.c:346:
808B11CDB37F0000:error:0188000A:bignum routines:test_reason_strings:DSA lib:test/errtest.c:347:
808B11CDB37F0000:error:0188000B:bignum routines:test_reason_strings:X509 lib:test/errtest.c:348:
808B11CDB37F0000:error:0188000D:bignum routines:test_reason_strings:ASN1 lib:test/errtest.c:350:
808B11CDB37F0000:error:0188000F:bignum routines:test_reason_strings:reason(15):test/errtest.c:351:
808B11CDB37F0000:error:01880010:bignum routines:test_reason_strings:EC lib:test/errtest.c:352:
808B11CDB37F0000:error:01880020:bignum routines:test_reason_strings:BIO lib:test/errtest.c:353:
808B11CDB37F0000:error:01880021:bignum routines:test_reason_strings:PKCS7 lib:test/errtest.c:354:
808B11CDB37F0000:error:01880022:bignum routines:test_reason_strings:X509V3 lib:test/errtest.c:355:
808B11CDB37F0000:error:01880026:bignum routines:test_reason_strings:ENGINE lib:test/errtest.c:356:
808B11CDB37F0000:error:01880028:bignum routines:test_reason_strings:UI lib:test/errtest.c:357:
808B11CDB37F0000:error:0188002A:bignum routines:test_reason_strings:ECDSA lib:test/errtest.c:358:
808B11CDB37F0000:error:0188002C:bignum routines:test_reason_strings:STORE lib:test/errtest.c:359:
808B11CDB37F0000:error:0188003C:bignum routines:test_reason_strings:reason(60):test/errtest.c:361:
808B11CDB37F0000:error:018C0000:bignum routines:test_reason_strings:fatal:test/errtest.c:362:
808B11CDB37F0000:error:018C0100:bignum routines:test_reason_strings:malloc failure:test/errtest.c:363:
808B11CDB37F0000:error:018C0101:bignum routines:test_reason_strings:called a function you should not call:test/errtest.c:364:
808B11CDB37F0000:error:018C0102:bignum routines:test_reason_strings:passed a null parameter:test/errtest.c:365:
808B11CDB37F0000:error:018C0103:bignum routines:test_reason_strings:internal error:test/errtest.c:366:
808B11CDB37F0000:error:018C0104:bignum routines:test_reason_strings:called a function that was disabled at compile-time:test/errtest.c:367:
808B11CDB37F0000:error:018C0105:bignum routines:test_reason_strings:init fail:test/errtest.c:368:
808B11CDB37F0000:error:01880106:bignum routines:test_reason_strings:reason(262):test/errtest.c:369:
808B11CDB37F0000:error:018C0107:bignum routines:test_reason_strings:operation fail:test/errtest.c:371:
808B11CDB37F0000:error:018C0108:bignum routines:test_reason_strings:invalid provider functions:test/errtest.c:372:
808B11CDB37F0000:error:01880109:bignum routines:test_reason_strings:interrupted or cancelled:test/errtest.c:373:
808B11CDB37F0000:error:0188010A:bignum routines:test_reason_strings:nested asn1 error:test/errtest.c:374:
808B11CDB37F0000:error:0188010B:bignum routines:test_reason_strings:missing asn1 eos:test/errtest.c:375:
808B11CDB37F0000:error:0188010C:bignum routines:test_reason_strings:unsupported:test/errtest.c:376:
808B11CDB37F0000:error:0188010D:bignum routines:test_reason_strings:fetch failed:test/errtest.c:377:
808B11CDB37F0000:error:0188010E:bignum routines:test_reason_strings:invalid property definition:test/errtest.c:378:
808B11CDB37F0000:error:018C010F:bignum routines:test_reason_strings:unable to get read lock:test/errtest.c:379:
808B11CDB37F0000:error:018C0110:bignum routines:test_reason_strings:unable to get write lock:test/errtest.c:380:
```

**Review by levitte — APPROVED** [review]

**Comment by levitte** [comment]
Thanks for your patience

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged to master and 3.0 - thanks @paulidale , @t8m and @levitte.
For 1.1.1 see #17069.
